### PR TITLE
Use Results class for WaterBridge analysis

### LINKED ISSUE #3270: update hydrogenbonds.WaterBridgeAnalysis to use .results
Follow the new AnalysisBase API and store results in hydrogenbonds.WaterBridgeAnalysis in `.results` instead of class attributes.

See omnibus issue #3261 for beackground and details. See PR #3263 as an example (DensityAnalysis)
- [ ] put all generated data into `self.results` instead of `self`
- [ ] keep managed attributes in place with deprecation message
- [ ] test deprecation messages/old attribute access
- [ ] update CHANGELOG: one entry in Changes (use .results) and one entry in Deprecations (attribute access to be removed in 3.0.0)

### PR DESCRIPTION
**Pull request by RMeli**
Fixes #3270

Changes made in this Pull Request:
- [x] `results.network`
- [x] `results.table`
- [x] `results.timeseries`

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/CHANGELOG
Comment (orbeckst) [reply package/CHANGELOG:244]: removes -> removed

File: package/CHANGELOG
Comment (orbeckst) [inline package/CHANGELOG:247]: removes -> removed

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:709]: Let's not include `results.table`

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:713]: Not an alias.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:717]: generate the table with `generate_table()`.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1560]: Should be a result – build it in `_conclude`.

@xiki-tempula how long does it take to build the timeseries? Should there be a ProgressBar showing progress (using verbose from `run()`?)

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1823]: Leave it to be `self.table`.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1773]: add a deprecation
```reST
.. versionchanged:: 2.0.0
   Return the generated table (as well as storing it as :attr:`table`).

.. deprecated:: 2.0.0
   In release 3.0.0, `generate_table()` will _only_ return the table and no longer store it in :attr:`table`.
```

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1768]: add explanations
```reST
Parameters
----------
output_format : {'sele1_sele2', 'donor_acceptor'}
     The output format of the `table` can be changed a fashion similar to
     :attr:`WaterBridgeAnalysis.results.timeseries` by changing the labels
     of the columns of the participating atoms.

Returns
-------
table : numpy.recarray
     A "tidy" table with one hydrogen bond per row, labeled according to 
    `output_format` and containing information of atom_1, atom_2, distance,
    and angle.
```

File: package/CHANGELOG
Comment (orbeckst) [inline package/CHANGELOG:180]: add that generate_table() returns the table

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1768]: If there's an explanation of the format of table then it could go here, at least in a short version.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1823]:
```suggestion
        self.table = table   # remove in 3.0.0
        return table
```

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1831]: remove — I don't think that we normally wrap attributes that are to be removed. Or do we, @IAlibay ?

Otherwise you need to store `self.table` in `self._table` and then use the property that you have here.

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:717]: also test that it *returns* the table

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:772]: remove (or test deprecation of `table`, in case we do that)

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (xiki-tempula) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:695]: I think one doesn't need to document a private variable.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1560]: It is always instant.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (xiki-tempula) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:693]: Sorry if I have missed something. I wonder if I can see the discussion with regard to `results.hbonds`?

**Review by xiki-tempula — CHANGES_REQUESTED** [review]
Thanks for the work. A few questions.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1560]: Because building timeseries is fast, we should just generate it in `_conclude` and assign it to `results.timeseries`. Then add the usual property with deprecation notice to `self.timeseries`.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:695]: Fair enough, then we also do not need to have a property alias for it.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:693]: This is a typo. @RMeli please change hbonds to network.
```suggestion
         :attr:`results.network` instead.
```

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (xiki-tempula) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1834]: I think we should just remove _network as it is a private variable and is not used outside the module.

Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1834]: Ok, that makes it even easier.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (RMeli) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:125]: Yes, once I change `timeseries` to `results.timeseries` I will change all the documentation accordingly.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (RMeli) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:695]: Good point. I was initially a bit confused because the `_network` was the attribute actually being used in the tests, therefore I thought it was something users could use as well (I often look at tests to see how things are used). Since it is private and there is a public counterpart, I will just remove this and deprecate `network` in favor of `results.network`.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (RMeli) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1834]: As mentioned above it was used a lot in the tests, but this PR will change that and therefore it is definitely better to just remove it.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (RMeli) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1773]: Great suggestion, thanks!

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (RMeli) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py]: Yep, typo (or, more likely, copy-paste gone wrong...). Thanks for spotting this!

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (RMeli) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py]: Thanks @xiki-tempula , I did not spotted the second one.

**Review by xiki-tempula — APPROVED** [review]
LGTM. Thanks for the work.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py]: Table will not be made a `results.table`. Instead we want to completely remove it for 3.0.0 (so that users will just use `generate_table()` to create it when they need it.

If you think we should issue a deprecation warning for the use of the attribute then we can do it. (I would then create `self._results` and have the table property just return that so that we don't trigger the warning when we set table in `generate_table()`.) 

Alternatively, we just put in a deprecation in the docs/CHANGELOG and be done.

**Review by orbeckst — CHANGES_REQUESTED** [review]
minor doc issues

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py]: > Alternatively, we just put in a deprecation in the docs/CHANGELOG and be done.
Sounds sensible to me.

**Comment by IAlibay** [comment]
Looks like #2942 will fix the current failures seen here @orbeckst, I'll leave it up to you if you want to wait until that is merged or just merge this one ahead of that.

(I do note that test coverage is somewhat low on the diff here, but I guess we can deal with this in a separate PR).

**Comment by orbeckst** [comment]
I checked the diff coverage https://app.codecov.io/gh/MDAnalysis/mdanalysis/compare/3287/diff  and don't see a problem. Happy to merge.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:679]: `timesteps` is used for generating the result table, which is in the format of
`"time", "donor_index" ...`

**Pull request merged**
